server/src/main/java/org/elasticsearch/cluster/coordination/Publication.java

```diff
@@ -0,0 +1,376 @@
+        private PublicationTargetState state = PublicationTargetState.NOT_STARTED;
+
+        public void setState(PublicationTargetState newState) {
+            switch (newState) {
+                case NOT_STARTED:
+                    assert false : state + " -> " + newState;
+                    break;
+                case SENT_PUBLISH_REQUEST:
+                    assert state == PublicationTargetState.NOT_STARTED : state + " -> " + newState;
+                    break;
+                case WAITING_FOR_QUORUM:
+                    assert state == PublicationTargetState.SENT_PUBLISH_REQUEST : state + " -> " + newState;
+                    break;
+                case SENT_APPLY_COMMIT:
+                    assert state == PublicationTargetState.WAITING_FOR_QUORUM : state + " -> " + newState;
+                    break;
+                case APPLIED_COMMIT:
+                    assert state == PublicationTargetState.SENT_APPLY_COMMIT : state + " -> " + newState;
+                    break;
+                case FAILED:
+                    assert state != PublicationTargetState.APPLIED_COMMIT : state + " -> " + newState;
+                    break;
+            }
+            state = newState;
+        }
+
+        public boolean isActive() {
+            return state != PublicationTargetState.FAILED
+                && state != PublicationTargetState.APPLIED_COMMIT;
+        }
+
+        public boolean isWaitingForQuorum() {
+            return state == PublicationTargetState.WAITING_FOR_QUORUM;
+        }
+
+        public boolean mayCommitInFuture() {
+            return (state == PublicationTargetState.NOT_STARTED
+                || state == PublicationTargetState.SENT_PUBLISH_REQUEST
+                || state == PublicationTargetState.WAITING_FOR_QUORUM);
+        }
+
+        public boolean isFailed() {
+            return state == PublicationTargetState.FAILED;
+        }
+
+        @Override
+        public String toString() {
+            return state.toString();
+        }
+    }
+
+    private class PublicationTarget {
+        private final DiscoveryNode discoveryNode;
+        private final PublicationTargetStateMachine publicationTargetStateMachine = new PublicationTargetStateMachine();
+        private boolean ackIsPending = true;
+
+        private PublicationTarget(DiscoveryNode discoveryNode) {
+            this.discoveryNode = discoveryNode;
+        }
+
+        @Override
+        public String toString() {
+            return discoveryNode.getId();
+        }
+
+        public void sendPublishRequest() {
+            if (publicationTargetStateMachine.isFailed()) {
+                return;
+            }
+            publicationTargetStateMachine.setState(PublicationTargetState.SENT_PUBLISH_REQUEST);
+            Publication.this.sendPublishRequest(discoveryNode, publishRequest, new PublicationTarget.PublishResponseHandler());
+            // TODO Can this ^ fail with an exception? Target should be failed if so.
```

Comment:
Does this question have an answer? I _think_ any failures are passed to the response handler.

Reply:
I'm deferring this TODO until we integrate this with Legislator.